Fix escaping markdown by rendering plaintext

kyrias opened the pull request:
We still need to parse "plaintext" messages through the markdown
renderer so that escappes are rendered properly.

kyrias commented:
(This is a WIP at the moment, will take care of the errors)

kyrias commented:
Should be ready fro reviewing now!

kyrias commented:
Oh, and I just realized that I accidentally got c819b433a247ca502daf600c3bc763f138262f5a in there, but it's not directly related to this PR, it just fixes the old messageinput component to use the new markdown render function.  I can split that out into its own PR if you want.

kyrias commented:
Okay, this isn't quite ready for merging yet, there are some more functions that need to be nulled-out, eg `"` will be HTML entity encoded currently.

kyrias commented:
And now that is fixed!  This should probably be tested out for a little while by more than just me before this is merged, to try to catch any other possible edge cases.

Kegsay commented on src/Markdown.js:
```diff
@@ -56,23 +57,47 @@ export default class Markdown {
     }
 
-    toHTML() {
+    render(html) {
```
Please don't call this `render`, as it is confusing and can make readers mistake the `Markdown` class as a React class.

Kegsay replied:
Also, it is not at all clear that `html` is a `boolean`. Please document what it is supposed to do.

kyrias replied:
What do you want me to rename it to?  `toHTML` isn't really an apt name anymore with this.

Kegsay replied:
Given most of this function is branching based on this conditional, separate functions sounds excellent, as it will also make it clearer at the call site what you're actually doing `toHTML` or `toPlaintext` or whatever.

Kegsay commented on src/Markdown.js:
```diff
@@ -56,23 +57,47 @@ export default class Markdown {
-                real_paragraph.call(this, node, entering);
+        } else {
+            renderer.out = function(s) {
+                this.lit(s);
```
Uhh, what is this doing? No comments.

kyrias replied:
Ah, seems I only added the comment in the commit message.  Will fix that.  The normal `out` function runs the output through a sanitizer that will entity encode things like quotation marks, which is problematic when it happens multiple times.

Kegsay replied:
Thanks for the explanation. What's `lit(s)` doing though?

kyrias commented:
Unit tests for what exactly?  Just the issue I found (`"` being entity encoded), or in general?  Because for the latter I could really use some suggestions for what to test.

Kegsay commented:
> Unit tests for what exactly? Just the issue I found (" being entity encoded), or in general? Because for the latter I could really use some suggestions for what to test.

The point is to test that the `Markdown` class actually works. We've already got a good list of stuff we've already found as bugs:
 - `"` being entity encoded.
 - Parsing escapes correctly if not markdown (e.g the test data in https://github.com/vector-im/riot-web/issues/2870 )
 - `-_-` not doing `<hr>` : https://github.com/matrix-org/matrix-react-sdk/pull/575#issuecomment-263729863
 - Check `<s>` works as intended : https://github.com/matrix-org/matrix-react-sdk/pull/575#issuecomment-263876049
 - Check `~~~` does not strikethrough.
 - Check that "plain" markdown doesn't incorrectly fail the `isPlainText()` check (e.g single paragraph test, multiple paragraph test)

kyrias commented:
Split the functions and added some test cases now.  It seems like while commonmark doesn't strip `<s>` tags, riot-web seems to not render them.  `<del>` tags work fine though.

Kegsay commented on src/Markdown.js:
```diff
@@ -56,12 +59,10 @@ export default class Markdown {
 
-    toHTML() {
-        const parser = new commonmark.Parser();
+    toHTML(html) {
```
You don't appear to use `html`.

kyrias replied:
Aah, relics that I missed in my `git add -p`!

Kegsay commented on src/components/views/rooms/MessageComposerInput.js:
```diff
@@ -523,8 +523,10 @@ export default class MessageComposerInput extends React.Component {
+            if (md.isPlainText()) {
+                contentText = md.toPlaintext();
+            } else {
+                contentHTML = md.toHTML(true);
```
Likewise I don't think you need the `true` anymore.

Kegsay commented on src/components/views/rooms/MessageComposerInputOld.js:
```diff
@@ -331,6 +331,7 @@ module.exports = React.createClass({
                 MatrixClientPeg.get().sendHtmlMessage(this.props.room.roomId, contentText, htmlText);
         }
         else {
+            const contentText = mdown.toPlaintext(false);
```
And here.

Kegsay commented on test/components/views/rooms/MessageComposerInput-test.js:
```diff
@@ -158,4 +158,85 @@ describe('MessageComposerInput', () => {
+        mci.handleReturn(sinon.stub());
+
+        expect(spy.calledOnce).toEqual(true);
+        expect(spy.args[0][1]).toEqual('Lorem ipsum dolor sit amet, consectetur adipiscing elit.');
```
Don't you need to assert that `spy.args[0][2]` is actually blank to be sure it didn't try to send this as HTML?

kyrias replied:
If it tried to send it as HTML the calledOnce assertion will fail, because then `sendHtmlMessage` will have been used rather than `sendTextMessage`

Kegsay replied:
Aha! Gotcha.

Kegsay commented on test/components/views/rooms/MessageComposerInput-test.js:
```diff
@@ -158,4 +158,85 @@ describe('MessageComposerInput', () => {
+        mci.handleReturn(sinon.stub());
+
+        expect(spy.calledOnce).toEqual(true);
+        expect(spy.args[0][1]).toEqual('Lorem ipsum dolor sit amet, consectetur adipiscing elit.\n\nFusce congue sapien sed neque molestie volutpat.');
```
Same here.

Kegsay commented:
LGTM!
